### DISCUSSION TITLE
Reading CGNS files in trame

### DISCUSSION THREAD

**Discussion by wyattnapier** [Kitware/trame#530] [Trame/VTK] [answered]
Hi! I am trying to read CGNS files in trame rather than a vtk or vtu file. I followed the same steps as were listed in the trame-tutorial and I copied most of my code from 04_application/solution.py. However, I modified the import statement such that it now is `from vtkmodules.vtkIOCGNS import vtkCGNSReader`. Then, I also changed the read line so that reader is defined as follows: `reader = vtk.vtkCGNSReader()`. However, I am still just getting the error: `ModuleNotFoundError: No module named 'vtkmodules.vtkIOCGNS'`. 

I also tried another implementation where `reader = vtk.vtkCGNSReader()`, but it results in an error as well: `AttributeError: module 'trame.widgets.vtk' has no attribute 'vtkCGNSReader'`.

How should I go about fixing this? I am restricted to CGNS files by the nature of my work.

**Comment by jourdain** [comment]
you may want to do `from trame.widgets import vtk as vtk_widgets` so you don't confuse the global `vtk` package with `trame.widgets.vtk` one.

**Reply by wyattnapier** [reply]
Hi! Thank you for such a quick response. Unfortunately, that didn't fix the issue.
![Screenshot 2024-05-30 161234](https://github.com/Kitware/trame/assets/117198964/33f2c1cd-7f0f-42e9-9c11-ee99c11734d7)

**Reply by jourdain** [reply]
My suggestion was to fix the following

```python
import vtk
reader = vtk.vtkCGNSReader()
```

**Reply by jourdain** [reply, marked as answer]
Otherwise looking at the [module name](https://github.com/Kitware/VTK/blob/master/IO/CGNS/vtk.module#L4C3-L4C18) it should be

```python
from vtkmodules.vtkIOCGNSReader import vtkCGNSReader
```

**Reply by wyattnapier** [reply]
Oh, I'm sorry, I guess I misunderstood. The last thing you said worked, so now I'm on to dealing with other issues. Thank you!

**Reply by jmythms** [reply]
@wyattnapier Did you ever get it to work? If yes, can you share the code?

**Reply by wyattnapier** [reply]
It was something like this:
```
from vtkmodules.vtkIOCGNSReader import vtkCGNSReader  
CURRENT_DIRECTORY = os.path.abspath(os.path.dirname(__file__))  
reader = vtkCGNSReader()  
reader.SetFileName(os.path.join(CURRENT_DIRECTORY, "pathname"))  
reader.Update()  
```
When you go to the documentation page, there is usually a link at the top that is prefaced by "#include" and then in that code you can typically find the module you need to import within such as vtkIOCGNSReader which is where vtkCGNSReader is defined and thus that is why I imported it like that. Additionally, you'll often be suggested to do something like vtk.vtkCGNSReader(), but because in my code I have the line, `from trame.widgets import vtk, vuetify, trame` it messes things up and you will just want to import the vtkCGNSReader as I did above and then just call it like I did before without the "vtk." before it. Let me know if you have any more questions.

**Reply by jmythms** [reply]
May I ask if you ended up using a specific Mapper for the cgns reader?

**Reply by wyattnapier** [reply]
I'm not sure if my approach was the best way to do it, but the output of the CGNS reader is a multi-block dataset so I had a nested for loop to iterate over the blocks (there is an outer block and more inner blocks within it). From there, I appended each inner block to a list. Then when I continued my pipeline I did it for each block so I ended up with actors for each block. I based most of my coding on [this trame application tutorial](https://kitware.github.io/trame/guide/tutorial/application.html#vtk_pipeline-id).

**Reply by jmythms** [reply]
Thank you!

**Comment by wyattnapier** [comment]
Hi @jourdain! I had one more question about CGNSReaders. I was hoping to upload a file by following this example: [https://github.com/Kitware/trame/blob/master/examples/00_howdoi/upload.py](upload.py). However, if I want to then have that file be the input to my vtkCGNSReader, I have been defining it with the line: `reader.SetFileName(os.path.join(CURRENT_DIRECTORY, "pathname"))`. I tried to copy this implementation below.


```
@state.change("file_upload")
def file_uploaded(file_upload, **kwargs):
    if file_upload is None:
        return
    file = ClientFile(file_upload)
    file_name = file_upload.get("name")
    ABS_PATH = os.path.abspath(file_name)
    print("absolute path:", ABS_PATH)

    reader = vtkCGNSReader()
    reader.SetFileName(ABS_PATH)
    reader.Update()
    print(reader.GetOutput().GetNumberOfBlocks())
```
Is there a way that I can pass the file_upload into the vtkCGNSReader?

**Reply by jourdain** [reply]
yes, you need to write it to disk so you can read from it.
